storage/storage.go

```diff
@@ -13,3 +20,120 @@ func NewStorage(bucket pail.Bucket) Storage {
 		bucket: bucket,
 	}
 }
+
+func parseName(name string) (start time.Time, end time.Time, numLines int64, err error) {
+	nameParts := strings.Split(name, "_")
+	startNanos, err := strconv.ParseInt(nameParts[0], 10, 64)
+	if err != nil {
+		return
+	}
+	start = time.Unix(0, startNanos)
+
+	endNanos, err := strconv.ParseInt(nameParts[1], 10, 64)
+	if err != nil {
+		return
+	}
+	end = time.Unix(0, endNanos)
+
+	numLines, err = strconv.ParseInt(nameParts[2], 10, 64)
+	if err != nil {
+		return
+	}
+	return
+}
+
+func (storage *Storage) getAllChunks(context context.Context, buildId string) ([]LogChunkInfo, error) {
+	buildPrefix := fmt.Sprintf("/%s/", buildId)
+
+	iterator, listErr := storage.bucket.List(context, buildPrefix)
+	buildChunks := []LogChunkInfo{}
+	if listErr != nil {
+		return nil, listErr
+	}
+	for iterator.Next(context) {
+		if strings.HasSuffix(iterator.Item().Name(), "metadata.json") {
+			continue
+		}
+		if strings.Contains(iterator.Item().Name(), "/tests/") {
+			keyParts := strings.Split(iterator.Item().Name(), "/")
+			buildId := keyParts[1]
+			test_id := keyParts[3]
+			name := keyParts[4]
+			start, end, numLines, nameErr := parseName(name)
+			if nameErr != nil {
+				return nil, nameErr
+			}
+			buildChunks = append(buildChunks, LogChunkInfo{
+				BuildId:  buildId,
+				TestId:   test_id,
+				Start:    start,
+				End:      end,
+				NumLines: int(numLines),
+			})
+		} else {
+			keyParts := strings.Split(iterator.Item().Name(), "/")
+			buildId := keyParts[1]
+			name := keyParts[2]
+			start, end, numLines, nameErr := parseName(name)
+			if nameErr != nil {
+				return nil, nameErr
+			}
+			buildChunks = append(buildChunks, LogChunkInfo{
+				BuildId:  buildId,
+				TestId:   "",
+				Start:    start,
+				End:      end,
+				NumLines: int(numLines),
+			})
+		}
+	}
+	return buildChunks, nil
+}
+
+func (storage *Storage) GetTestLogLines(context context.Context, buildId string, testId string) (LogIterator, error) {
+	chunks, err := storage.getAllChunks(context, buildId)
+	if err != nil {
+		return nil, err
+	}
+
+	testChunks := []LogChunkInfo{}
+	for i := 0; i < len(chunks); i++ {
+		// Find our test id
+		if chunks[i].TestId == testId {
+			testChunks = append(testChunks, chunks[i])
+		}
+	}
+	sort.Slice(testChunks, func(i, j int) bool {
+		return testChunks[i].Start.Before(testChunks[j].Start)
+	})
+
+	testTimeRange := TimeRange{
+		StartAt: testChunks[0].Start,
+		EndAt:   testChunks[len(testChunks)-1].End,
+	}
+
+	testChunkIterator := NewBatchedLogIterator(storage.bucket, testChunks, 4, testTimeRange)
```

Comment:
why is it better to use the batched iterator as opposed to the parallelized iterator?

Reply:
BatchLogIterator and the parallelized iterator are actually the same implementation, but the parallelized iterator downloads every chunk, while the batch iterator does the next "n" chunks based on the batch size once it has exhausted existing chunks. The ParallelizedLogIterator is just a BatchedLogIterator where n = len(chunks).

We'll presumably want to tune this for performance. If we're ok just always loading everything into memory than the ParallelizedLogIterator is probably faster. It's a little more complicated than that since neither one actually fully downloads the log chunks from the bucket. They both just open the HTTP connection and get the initial HTTP response using the .Get() method from pail, but I'm pretty sure that doesn't actually read in the whole body into memory.

Reply:
That sounds good to me.

I'd always assumed it was buffered somewhere on the client until we read from the reader. Is there a way for the client to throttle the server so it only sends when we're ready?

Reply:
I _think_ if you don't call read() what will happen is that the TCP buffer will fill up and the server will wait to send more bytes. I feel like I'd need to test this empirically though to be sure. It could be the case that ParalleizedLogIterator is fine in that case. It's an easy change to make any time though.

Reply:
At a glance it looks like the TCP buffer is calculated using this formula: ```max(87380, min(4 MB, tcp_mem[1]*PAGE_SIZE/128))
```
In principle unless the go runtime is calling recv for us then that's around memory will be used until we read from the body.

Reply:
I'm pretty sure in practice that means 4 MB.

Reply:
See https://github.com/johndaniels/gobuffertest